components/ai_chat/renderer/page_text_distilling.cc

```diff
@@ -204,4 +230,89 @@ void DistillPageText(
+  // our extractor script. Because the injector-script is wrapped in an IIFE, we
+  // leverage custom events dispatched against the window object to communicate
+  // between the two scripts.
+  //
+  // The extractor script is currently hard-coded to request content at level 3.
+  // Level 3 is the "FULL" level, which is the highest content level. The level
+  // could be tied to the currently-active model (e.g., smaller models could be
+  // defaulted to a smaller level).
+  std::string script =
+      script_content + absl::StrFormat(
+                           R"(
+      new Promise(function(resolve, reject) {
+          window.addEventListener('%s', (event) => {
+              if (event?.detail?.result) {
+                  return resolve(event.detail.result);
+              }
+              reject("No result in %s event");
+          });
+      });
+
+      window.dispatchEvent(new CustomEvent(
+          '%s', { detail: { level: 3 } }
+      ));
+    )",
+                           "LEO_DISTILL_RESULT", "LEO_DISTILL_RESULT",
+                           "LEO_DISTILL_REQUESTED");
+
+  blink::WebScriptSource source =
+      blink::WebScriptSource(blink::WebString::FromUTF8(script));
+
+  auto on_script_executed =
+      [](base::OnceCallback<void(const std::optional<std::string>&)> callback,
+         std::optional<base::Value> value, base::TimeTicks start_time) {
+        if (value && value->is_string() && !value->GetString().empty()) {
+          std::move(callback).Run(value->GetString());
+          return;
+        }
+
+        std::move(callback).Run({});
+      };
+
+  // Execute the combined script as a single source
+  render_frame->GetWebFrame()->RequestExecuteScript(
+      world_id, base::span_from_ref(source),
+      blink::mojom::UserActivationOption::kDoNotActivate,
+      blink::mojom::EvaluationTiming::kAsynchronous,
+      blink::mojom::LoadEventBlockingOption::kDoNotBlock,
+      base::BindOnce(on_script_executed, std::move(callback)),
+      blink::BackForwardCacheAware::kAllow,
+      blink::mojom::WantResultOption::kWantResult,
+      // Because we are using a promise to resolve the result, we will use the
+      // `kAwait` option to ensure the promise is resolved before the callback
+      // is invoked.
+      blink::mojom::PromiseResultOption::kAwait);
+}
+
+std::optional<std::pair<std::string, bool>> LoadSiteScriptForHost(
+    std::string_view host) {
+  static const std::map<std::string, std::pair<int, bool>>
+      kHostToScriptResource = {
+          {"github.com",
+           {IDR_AI_CHAT_DISTILLER_SCRIPTS_GITHUB_COM_BUNDLE_JS, false}},
+          {"x.com", {IDR_AI_CHAT_DISTILLER_SCRIPTS_X_COM_BUNDLE_JS, true}},
+      };
+
+  auto it = kHostToScriptResource.find(
+      net::registry_controlled_domains::GetDomainAndRegistry(
+          host, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES));
+
+  if (it != kHostToScriptResource.end()) {
```

Comment:
In general it is better to have the unhappy path as an early return, and the happy path as the straight one, as the compiler can restrict RVO for the object you're creating below.